### ISSUE TITLE
Feature request: MIDI sound-font (.sf2) support for samples/instruments

### ISSUE THREAD

**Issue by goofyseeker311** [schismtracker/schismtracker#310]
Currently it seems .sf2 midi sound font files are not supported, I assume MIDI is only for inputting notes not actually being the sample-instrument source. This feature would be nice.

**Comment by mrpapersonic** [comment]
You can import samples from SF2 files now with 7e28360b6ebdd7bfee80b96b47eeaee3ccc21282 (with a limit of 236 samples), but nothing "integrating" .mid and .sf2 files. midi and trackers [generally don't mix together well anyway](https://i.imgur.com/XDBOLWD.jpg).

**Comment by goofyseeker311** [comment]
I thought midi files would be similar to the tracker files. the sf2 samples feature is the first priority and anything beyond is secondary tho. mostly because midi is used already as input, but only from the input method entry side, not the samples side.

edit: being a midi tracker alongside standard tracker would be neat tho.

edit 2: samples not being stereo should not be too great of an issue. however sf2 does that tho.

edit 3: I think midi and trackers are not too different in the end. it could work just fine. midi sf2 samples import is a nice step and option.

**Comment by goofyseeker311** [comment]
you sure the sf2 file does not just plain simple contain two channels of the sample?

whatever this contains: https://www.synthfont.com/sfspec24.pdf

**Comment by mrpapersonic** [comment]
yes, but they are not guaranteed to be next to each other in the file, in order of left channel then right channel (split stereo), which makes it a pain to implement with the current code

**Comment by mrpapersonic** [comment]
note: it could probably be viable to make two temporary samples and then merge them... but that would be a waste of memory and have a lot of overhead.

Maybe the best way would be creating an sf2-specific function that takes offsets for left channel and right channel.

**Comment by goofyseeker311** [comment]
ok I have now tested briefly blindly (no manuals or tutorials read) the sf2 samples/instruments loading on version [6616689](https://github.com/schismtracker/schismtracker/commit/6616689013a4151ec8f7f2473949d67669eba70c) from the github actions [amd86_64](https://github.com/schismtracker/schismtracker/actions/runs/14628038257#:~:text=schismtracker%2D20250423%2Dmingw32%2Dx86_64).
a random .sf2 file loads up nicely from the f4 instruments list menu (not from the shift-f4 instrument library menu).
then the samples are showing and playable on the f3 sample list menu.

how would you get the samples back as channel instruments tho?
I can open the impulse tracker defaults songs, they play nice.
additional notes can be added in the f2 pattern editor for the sample songs.
but when I try to add the sf2 sample instruments (?) nothing happens.
f5 does not even start to play back cycle the pattern.

so what I'm doing wrong from doing an empty it file sf2 import usage?
yes in the new user way wrong, as in idk what I'm supposed to do to get the working samples as patternable instruments.

edit: I suppose that the sf2 pack should not be loaded from the f4 instrument list display, now the sf2 lib sits there as the first instrument. but somehow the shift-f4 menu loading of the sf2 lib does not seem to work. so whats the right way?
stated in another words: how do I assign the sample list back to the instrument list as usable instruments?
instrument line enter press always opens a file dialog, instead of letting to pick from the sample list.

note: sf2 sound library files are a super simple way to get some music making going. like using some roland synth instruments.

edit: shift-f3 does not work in stated windows build. only the window menu samples -> sample library works.

tldr: you should not be able to load an .sf2 file as an instrument library, only as a sample library. now it works like it should.
when the .sf2 is as a sample library, then you can pick to the sample list from the sample library, then it ask if it should add a corresponding instrument.

edit2: f5 related song length should be the whole file all patterns by default, now it seems to be nothing by default. f6 works with a clean file but f5 does not. f5 length should be specifically shortened if required. well in another words, all any patterns should be played in order with a clean file using f5. ie if no pattern order then loop all patterns. yep f12 modification fixes that but defaults should be the best.

edit: I assume instruments are tracker internal instruments, and samples are tracker external samples, or should be.
yep because I dont try the conventional way then the unobvious buggy ways come into the light.

**Comment by goofyseeker311** [comment]
other bugs:

keyboard press: alt-gr+? = \ does not work for path writing.
directories settings are not navigable dialog.
you cannot copypaste from windows clipboard such path strings.

edit: also shift-f4 to instrument library does not work in windows. but the window menu mouse click works.

**Comment by mrpapersonic** [comment]
> keyboard press: alt-gr+? = \ does not work for path writing.

pretty much nothing outside of the US locale/keyboard is tested (so literally anything w/ alt-gr basically will not work, because its a pain and it's very cumbersome for me to actually test it without any foreign keyboards available).

for the record, a regular slash works perfectly as a path separator on Windows. (try it!)

> directories settings are not navigable dialog.

what?

> you cannot copypaste from windows clipboard such path strings.

copy/paste is only implemented for the message editor and a few other things (IIRC) but its on my todo to add it in other places

> edit: also shift-f4 to instrument library does not work in windows. but the window menu mouse click works.

that keybind is ctrl-f4, not shift-f4

> edit: shift-f3 does not work in stated windows build. only the window menu samples -> sample library works.

ditto, this is ctrl-f3

> edit2: f5 related song length should be the whole file all patterns by default, now it seems to be nothing by default. f6 works with a clean file but f5 does not. f5 length should be specifically shortened if required. well in another words, all any patterns should be played in order with a clean file using f5. ie if no pattern order then loop all patterns.

No, and this will never be implemented; the *proper* way to do it is define which patterns you want to play in the orderlist (F11). Implicitly editing (and possibly corrupting if the code is not absolutely 100% correct) of the orderlist is extremely annoying IMO.

**Comment by goofyseeker311** [comment]
"directories settings are not navigable dialog."

you cannot press enter in the directories settings to browse select a directory. you have to write a very long directory path by hand. in windows specifically. the normal starting location is c:\ which is far away from the user document folder, which is like c:\users\username\Documents and even longer with any additions. thats why copypaste would be a secondary avoidance of typing very long paths by hands. windows is not based on the drive root, but in the user folders for user data. "c:\\" is an old dos convention that does not apply very well to modern win11.

why the ctrl-f4 says shift-f4 etc then in the corresponding menu? also the windows menu shortcuts says its shift-f4 etc instead of ctrl-f4.

f5 logic is only one if statement to check (if pattern list is empty play f6 even if f5 was pressed). f5 does nothing currently possibly.

windows default directory separator is \ not /. the default paths in the settings have \ not / in them. currently even "c:\\".
how about using some default gui input element to read any input, instead of binding directly to the key input codes.
ie sdl2 should have an operating system independent input box or just generic input reading not tied to the os specifics.

edit: user experience vs programming logic correctness.

I'm also a programmer, so the \ or / thing in windows did not escape me. but still inconvenient to not being able to use os local input conventions.

quick googling of generic text input for sdl2 gives this:
https://wiki.libsdl.org/SDL2/Tutorials-TextInput

this would provide you with the user characters directly without needing to manually process user key presses (alt-gr + ? would be \ directly). ie user language to input string conversion automatically.

if you dont want to deal with the free-text input complications then just provide the folder browse selection dialog, like in the instruments import file dialog. but a path string in any os is still an abstract thing so it would not create any complications. if user types something the os cant handle then the local os file handling logic will error in software. also SDL should have the freetext language fonts freetype image generators by default for any language automatically, so thats not an issue also.

my main rig is windows, so will not force myself to use linux conventions on windows. linux is for server stuff.
any user experience aware software should not force any other os conventions on the users.

**Comment by mrpapersonic** [comment]
> "directories settings are not navigable dialog."
> 
> you cannot press enter in the directories settings to browse select a directory.

What exactly are you suggesting? What does "browse select" mean?? You can recurse into directories just fine in all of the loading menus.

> you have to write a very long directory path by hand. in windows specifically. the normal starting location is c:\ which is far away from the user document folder, which is like c:\users\username\Documents and even longer with any additions. thats why copypaste would be a secondary avoidance of typing very long paths by hands. windows is not based on the drive root, but in the user folders for user data. "c:\" is an old dos convention that does not apply very well to modern win11.

really? I can browse and recurse into directories just fine..

The normal starting path is not `C:\`, but the user's Documents folder on the first startup of schism. I've implemented many fallbacks to make sure this works correctly even on the most outdated of systems (Windows 95). These can also be configured after in the Variables screen (F12)

> why the ctrl-f4 says shift-f4 etc then in the corresponding menu? also the windows menu shortcuts says its shift-f4 etc instead of ctrl-f4.

https://github.com/schismtracker/schismtracker/blob/6616689013a4151ec8f7f2473949d67669eba70c/sys/macosx/macosx-sdlmain.m#L340

It's been wrong since 2009 apparently and was a rollover from when the macosx code was ported.

> f5 logic is only one if statement to check (if pattern list is empty play f6 even if f5 was pressed). f5 does nothing currently possibly.

F5 plays the song, F6 plays the current pattern. they are two different things.

> windows default directory separator is \ not /. the default paths in the settings have \ not / in them. currently even "c:\". how about using some default gui input element to read any input, instead of binding directly to the key input codes. ie sdl2 should have an operating system independent input box or just generic input reading not tied to the os specifics.
> 
> edit: user experience vs programming logic correctness.
> 
> I'm also a programmer, so the \ or / thing in windows did not escape me. but still inconvenient to not being able to use os local input conventions.
> 
> quick googling of generic text input for sdl2 gives this: https://wiki.libsdl.org/SDL2/Tutorials-TextInput
> 
> this would provide you with the user characters directly without needing to manually process user key presses (alt-gr + ? would be \ directly). ie user language to input string conversion automatically.
> 
> if you dont want to deal with the free-text input complications then just provide the folder browse selection dialog, like in the instruments import file dialog. but a path string in any os is still an abstract thing so it would not create any complications. if user types something the os cant handle then the local os file handling logic will error in software. also SDL should have the freetext language fonts freetype image generators by default for any language automatically, so thats not an issue also.

We are already doing this. The amount of keybinds within schism makes it a pain to distinguish whether the user means to e.g. cut a sample loop or input an integral (Option+B on macosx). It doesn't help that schism treats AltGr as Alt by default and has so since like, 2007 or so.

Possibly a solution, would be to return a real value for whether a key was "handled" by a keydown handler. Then we could just send the input as text input to the already existing functions. That could possibly introduce more bugs though... :(

> my main rig is windows, so will not force myself to use linux conventions on windows. any user experience aware software should not force any other os conventions on the users.

I suggested a workaround, never did I "force [you] to use linux conventions on windows". This is obviously a bug, but it's one I can't very readily "fix".

> linux is for server stuff. 

[One in every 25 desktops is running linux](https://gs.statcounter.com/os-market-share/desktop/worldwide). Stop kidding yourself.

**Comment by goofyseeker311** [comment]
main point is that I dont want to write the directory path of the settings manually. nothing else.
you cannot browse those directories on windows.

I will not submit to linux forced behaviour, because linux is not mainstream on desktop for normal users.
why use / when it should not be used. thats only from linux, not from windows or even from dos/amiga, where ST/IT was original made. / forced is literally linux syntax forced. which you forced me to use, instead of wanting to fix the \\.

also f5/f8 is the default visible action in the ui, instead of the f6. not even visible in the ui.
new users (not elite linux expert users) only wysiwyg. linux experts expect people to know every command not wysiwyg. just saying.

I dont believe even 1/10 of desktops, but even then no reason to force linux practices on any app ui.

just friendly suggestions. but you do you. farewell.

**Comment by goofyseeker311** [comment]
additional current sf2 midi sounds implementation issues (two days ago or so, 20250526 github actions x86_64):

playing any samples or instruments causes a crackle at the start of the sample, using a generic hq midi sf2 sound font file.
ie like the sample was full on right away causing a ramp-up spike crackle. ie no lowpass filtering or similar.

**Comment by mrpapersonic** [comment]
could you provide the .sf2 file for testing purposes?

**Comment by goofyseeker311** [comment]
https://musical-artifacts.com/artifacts/5582

Cocoto HQ Soundfont by IzzBloxian

possibly its about the sf2 file, not about schismtracker.
have not tested with any other sf2 sound font yet, like some generic roland synth soundfont .sf2 file.

**Comment by goofyseeker311** [comment]
let me rephrase the issue.

if you play the instrument again while its playing, it will get the crackle.
the instrument will not crackle if you play the instrument first time or after it has stopped playing.
with long instruments like chorus its easy to detect.

I think it should play always as if its was on independent different output channel regarding the mixer.
actually the same fast repeat crackle issue is present in the impulse tracker default .it files instruments too.

so its not the .sf2 file issue, its the instrument play logic/mechanism in schismtracker.
its because the previous sample stops and transitions to the next (same) sample with a hard impulse (no pun intended).

there is no issue with looping samples internally because they are matched internally. scratch that.
not always the case, leading to the same crackle, say "lead guitar" from the blue flame impulse tracker song.
it crackles when it loops.

fix might be to automatically make a better filtered smoothed low-pass transition for all instruments.

**Comment by goofyseeker311** [comment]
sdl2 issue? cracking happens also when stopping song playback and exit. etc.

anyways this issue normally surfaces when editing sounds/samples, and trying to make sure they dont crackle when played.
if the cut point is where there is too much volume, when starting or stopping, it will crackle when played or looped.

automatic instrument handling would be better than manually needing to adjust a .sf2 file for such issue.
midi has default sample playback smoothing features? hifi issue, possibly but should be addressed for best quality.

edit: smoothing = automatic instrument fast fade-in/out to prevent crackles.
and/or channel instrument volume/spectrum matching/blending during fade-in/out, might work better for looping.

edit: any of these issues are on the latest github actions version, 20250530 mingw32 x86_64.

**Comment by mrpapersonic** [comment]
Not an SDL2 issue, this is known (I encountered it when messing around with midi some time ago)

I believe the issue is with `song_keydown_ex`: https://github.com/schismtracker/schismtracker/blob/953fcd39d9a3496e14d43d1e54d554b104d5c5d2/schism/audio_playback.c#L396 because I don't think any volume ramping is done here. Maybe the solution is as simple as `chan->flags |= CHN_VOLUMERAMP;` ?? and then filling in the volume ramp info.

Though ideally all of this crap should be in the player code anyway, and we shouldn't duplicate all the stuff in `csf_instrument_change`...

**Comment by goofyseeker311** [comment]
at least the fade-in, fade-out, and fade-out/in should be an option, if not default, to preserve the normal working so far. or fade-out/in should be the default always, and optionally can be turned off. fade-out/in would also work for the looping instruments, like song mixing fade-out/in by djs.

edit: its not only the player, its the instrument F3 menus where its most audible. if player code handles all instrument playing then there of course.

**Comment by goofyseeker311** [comment]
I'm certain that low-pass filtering all channels at the output sampling frequency (or 1/2 or double) before outputting them would solve all the crackle issues, as it removes all impulses caused by random sample insertion. I wonder if any filtering is done anywhere, either by schismtracker or sdl2 audio layers. its like when downsampling or upsampling you need to low pass filter to get the signal spectrum correct, removing any out of band frequency spikes. they taught this in uni signal processing class.

this is (should be) standard practice in all audio.

**Comment by mrpapersonic** [comment]
I am not an expert by any means in audio processing, but (from my understanding) the FIR filter interpolation option in the Preferences menu (Shift-F5) should by definition perform a low-pass filter.

For context, linear interpolation is selected by default for mainly historic and cultural reasons. Module files are encouraged to be very small (or rather, *modular*), and as such, many use very lo-fi, low-quality samples. Those samples usually will sound terrible under a hi-fi interpolation option, as used in modern DAWs. I even play most modules without interpolation enabled at all.

**Comment by goofyseeker311** [comment]
well ramp up time it is then. any way that consistently and reliable removes the impulse spike crackles. even checking if the current system does the filters.

edit: mixing mode in the shift-F5 is default 16-bit linear, not the fir filter. yes assumes that samples are "correct" and do not crackle by themselves.

edit: turning off interpolation with the cocoto sf2 file sounds absolutely horrible. fir filter sounds marginally better than the default linear but still crackles a ton. I'm definitely on the hifi side, not targeting small file size, but high output audio quality. my own use case is hq music production as output wave file (into mp3 or ogg, or aac audio in mp4), not always small sized .mod, .it or similar. both targets should be equally important use cases.

crackle = snap when instrument is played on top/during of another instrument, and at start and end of the instrument.

the ramp volume at start of sample: enabled setting is not yet doing the right thing (20250530 version).
if ramp setting is disabled, there is double crackle snap when repeating same instrument on top/during same instrument.
I guess the ramp setting currently tries to make the ramping effective but still crackle snaps.

edit: cocoto hq .sf2 file choralOohhHQR instrument is a great example of the crackle happening in all three cases, during, at start and at end. even using the fir filter.

simplest fix (assuming low-pass filter is already being properly used) is making the ramp-up sample time longer. ie no crackle snap happens because volume ramp-up is slow enough. maybe a ramp-up time setting next to the enabled/disabled setting?

anyways solving the crackling issue would increase the quality of even the low quality samples tracks playback, if the main issue is the samples playback start/end/during crackling.

**Comment by goofyseeker311** [comment]
main point about the cracking fix automatically is to avoid manual fixing of .sf2 and having tolerance of "not perfect" samples, being turned into usable samples automatically.

I assume that having the ramp-up time will affect the start timing of the actual sample, maybe to have 50% volume hit the actual beat timing? or something similar. for the cases of extended ramp-up times.

in all cases the resulting ramp-up/down time needs to be longer. to achieve a hq result. which is desired.

**Comment by goofyseeker311** [comment]
the sample initial channel open pop sound is not yet gone by any updates up to version 237868c.

**Comment by goofyseeker311** [comment]
as of version 3183824, there is still the snap crackle sound in the F3 sample menu while playing some samples as test.
the crackle sounds like old vinyl disc track player head random noise. all filters and ramping at maximum.

**Comment by mrpapersonic** [comment]
yes, because it hasn't been fixed yet.
